src/main/scala/com/microsoft/hyperspace/index/dataskipping/DataSkippingIndex.scala

```diff
@@ -0,0 +1,132 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.microsoft.hyperspace.index.dataskipping
+
+import org.apache.spark.sql._
+import org.apache.spark.sql.functions.input_file_name
+
+import com.microsoft.hyperspace.index._
+import com.microsoft.hyperspace.index.dataskipping.sketch.Sketch
+
+case class DataSkippingIndex(
+    sketches: Seq[Sketch],
+    sketchOffsets: Seq[Int],
+    override val properties: Map[String, String] = Map.empty)
+    extends Index {
+  assert(sketches.nonEmpty)
+  assert(sketches.length + 1 == sketchOffsets.length)
+
+  override def kind: String = DataSkippingIndex.kind
+
+  override def kindAbbr: String = DataSkippingIndex.kindAbbr
+
+  override def indexedColumns: Seq[String] = sketches.flatMap(_.columns)
+
+  override def referencedColumns: Seq[String] = indexedColumns
+
+  override def withNewProperties(newProperties: Map[String, String]): DataSkippingIndex = {
+    copy(properties = newProperties)
+  }
+
+  override def statistics(extended: Boolean = false): Map[String, String] = {
+    Map("sketches" -> sketches.mkString(", "))
+  }
+
+  override def canHandleDeletedFiles: Boolean = true
+
+  override def write(ctx: IndexerContext, indexData: DataFrame): Unit = {
+    indexData.write.mode(SaveMode.Overwrite).parquet(ctx.indexDataPath.toString)
+  }
+
+  override def optimize(ctx: IndexerContext, indexDataFilesToOptimize: Seq[FileInfo]): Unit = {
+    val indexData = ctx.spark.read.parquet(indexDataFilesToOptimize.map(_.name): _*)
+    write(ctx, indexData.repartition(1))
+  }
+
+  override def refreshIncremental(
+      ctx: IndexerContext,
+      appendedSourceData: Option[DataFrame],
+      deletedSourceDataFiles: Seq[FileInfo],
+      indexContent: Content): (Index, Index.UpdateMode) = {
+    if (appendedSourceData.nonEmpty) {
+      write(ctx, index(ctx, appendedSourceData.get))
+    }
+    (this, Index.UpdateMode.Merge)
+  }
+
+  override def refreshFull(
+      ctx: IndexerContext,
+      sourceData: DataFrame): (DataSkippingIndex, DataFrame) = {
+    val updatedIndex = copy(sketches = DataSkippingIndex.resolveSketch(sketches, sourceData))
+    (updatedIndex, updatedIndex.index(ctx, sourceData))
+  }
+
+  /**
+   * Creates index data for the given source data.
+   */
+  def index(ctx: IndexerContext, sourceData: DataFrame): DataFrame = {
+    val aggregateFunctions = sketches.flatMap { s =>
+      val aggrs = s.aggregateFunctions
+      assert(aggrs.nonEmpty && aggrs.length == s.numValues)
+      aggrs.map(aggr => aggr.as(DataSkippingIndex.normalizeColumnName(aggr.toString)))
+    }
+    val fileNameCol = "input_file_name"
+    val indexData = sourceData
+      .groupBy(input_file_name().as(fileNameCol))
```

Comment:
Question: 
is this scalable for millions of files? how many tasks will be created?
Could you add some benchmark number (for index creation) using 1TB/100KChunk dataset? (or bigger?)

Reply:
My guess - it won't be slower than creating covering index data.

Creating covering index data - each local node makes a vertical slice of the source data. Then those vertical slices need to be shuffled all around the cluster because of sorting and bucketing.

Creating data skipping index data - each local node does aggregation for the source data. Then those aggregated data needs to be shuffled all around the cluster, but this time the shuffle size will be much smaller.

I'd like to do the measurement, but last time I tried to create big data in Spark it was too slow. Do you have any tips on this?

Reply:
Results for TPC-H 1TB, 1K/10K/100K chunks, lineitem, index suitable for Q6, 199 executors
- CoveringIndex (indexed: l_shipdate, l_discount, l_quantity, included: l_extendedprice)
  - 1K chunks: 3m 05
  - 10K chunks: 3m 12
  - 100K chunks: 4m 54
- DataSkippingIndex (MinMaxSketch/ValueListSketch on l_shipdate/l_discount/l_quantity)
  - 1K chunks: 27 sec
  - 10K chunks: 54 sec
  - 100K chunks: 2m 24

Reply:
Thanks! The result looks great 👍 Seems already have the rule implementation for minmax/value list :)
Could you also share index creation time & index data size for both CoveringIndex & DataSkippingIndex?

Reply:
These are the index creation times, not query execution times. Sorry I didn't make it clear 😅

Data skipping indexes are only effective when the indexed columns have some correlation with the partitioning columns, and the data I used was unpartitioned, so DS cannot be used.

Index size was 31 GB for CoveringIndex and 4/13/150 KB for DataSkippingIndex for 1K/10K/100K chunks respectively.

Reply:
I don't have datasets with 1 million files. Since the index data creation code is fairly simple, I think the results for 1-100K files are enough.